### ISSUE TITLE
Support asciidoc format.

### ISSUE THREAD

**Issue by fidergo-stephane-gourichon** [asamuzaK/url2clipboard#29]
It would be nice to support Asciidoc markup links. Basically: 

http://asciidoctor.org[Asciidoctor]
link:++http://example.org/?q=[a b]++[URL with special characters]
link:http://example.org/?q=%5Ba%20b%5D[URL with special characters]

Ref: http://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#links

Cheers!

**Comment by asamuzaK** [comment]
http://asciidoctor.org - automatic!
http://asciidoctor.org[Asciidoctor]

Which is the popular way?
Should both be needed?

**Comment by fidergo-stephane-gourichon** [comment]
# Short answer

* If lazy, use `link:url[text]` syntax always.
* In **URL part**, escape (percent-hexa url-encoded) the following characters:

```
0x09 0x0A 0x0B 0x0C 0x0D 
0x20 space
0x3C and 0x3E < >
0x5B and 0x5D [ ]
0x5E ^ 
0x60 backtick `
0x7B 0x7C 0x7D {|}
```
* In URL also escape all chars that don't fit in utf-8 in 8 byte.
* In **text part** escape (by prefixing with a backslash): the closing bracket "]" 0x5D.

The rest below is only research, details and rationale.

# Basic cases

> http://asciidoctor.org - automatic!

That "plain URL" is a non-topic. If user wants this, they can copy from URL bar (c.g. Ctrl-L Ctrl-C).

> http://asciidoctor.org[Asciidoctor]

Than one is the regular link-with-text that is used any time the URL has no special character.

# Special characters

Those below show that special characters should be escaped.

```
link:++http://example.org/?q=[a b]++[URL with special characters]
link:http://example.org/?q=%5Ba%20b%5D[URL with special characters]
```

If you want to be always safe, you must strip or escape some characters in URL.

IMHO there are some characters (especially low-number, non-printable) most users won't be upset if these special characters are simply stripped. Multibyte characters are necessary, still. Escape them but don't strip them.

The command line in next section uses recode to generate all 8-bit characters and pandoc to convert a link with all those special characters into asciidoc. If you don't know pandoc, you should check it out: https://pandoc.org/

From the result below, I guess the set of characters to be escaped (percent-hexa url-encoded) in URL is:
```
0x09 0x0A 0x0B 0x0C 0x0D 
0x20 space
0x3C and 0x3E < >
0x5B and 0x5D [ ]
0x5E ^ 
0x60 backtick `
0x7B 0x7C 0x7D {|}
```
as well as all chars that don't fit in utf-8 in 8 byte.

From the result below, I guess the set of characters to be escaped (by appending a backslash) in link text is:

```
0x5D closing bracket
```


## Gory details

### Which characters to escape in URL

The output below provides hint to figure out.

```
{ echo -n '<a href="' ; recode -f </dev/null iso-8859-1/TEST8..utf-8 | sed 's/"//' ; echo '">b</a> end' ; } | pandoc  -i /dev/stdin -f html -t asciidoc | hd

00000000  00 01 02 03 04 05 06 07  08 25 32 30 25 32 30 25  |.........%20%20%|
00000010  30 41 25 30 42 25 30 43  25 30 41 0e 0f 10 11 12  |0A%0B%0C%0A.....|
00000020  13 14 15 16 17 18 19 1a  1b 1c 1d 1e 1f 25 32 30  |.............%20|
00000030  21 23 24 25 26 27 28 29  2a 2b 2c 2d 2e 2f 30 31  |!#$%&'()*+,-./01|
00000040  32 33 34 35 36 37 38 39  3a 3b 25 33 43 3d 25 33  |23456789:;%3C=%3|
00000050  45 3f 40 41 42 43 44 45  46 47 48 49 4a 4b 4c 4d  |E?@ABCDEFGHIJKLM|
00000060  4e 4f 50 51 52 53 54 55  56 57 58 59 5a 25 35 42  |NOPQRSTUVWXYZ%5B|
00000070  5c 25 35 44 25 35 45 5f  25 36 30 61 62 63 64 65  |\%5D%5E_%60abcde|
00000080  66 67 68 69 6a 6b 6c 6d  6e 6f 70 71 72 73 74 75  |fghijklmnopqrstu|
00000090  76 77 78 79 7a 25 37 42  25 37 43 25 37 44 7e 7f  |vwxyz%7B%7C%7D~.|
000000a0  c2 80 c2 81 c2 82 c2 83  c2 84 c2 85 c2 86 c2 87  |................|
000000b0  c2 88 c2 89 c2 8a c2 8b  c2 8c c2 8d c2 8e c2 8f  |................|
000000c0  c2 90 c2 91 c2 92 c2 93  c2 94 c2 95 c2 96 c2 97  |................|
000000d0  c2 98 c2 99 c2 9a c2 9b  c2 9c c2 9d c2 9e c2 9f  |................|
000000e0  25 43 32 25 41 30 c2 a1  c2 a2 c2 a3 c2 a4 c2 a5  |%C2%A0..........|
000000f0  c2 a6 c2 a7 c2 a8 c2 a9  c2 aa c2 ab c2 ac c2 ad  |................|
00000100  c2 ae c2 af c2 b0 c2 b1  c2 b2 c2 b3 c2 b4 c2 b5  |................|
00000110  c2 b6 c2 b7 c2 b8 c2 b9  c2 ba c2 bb c2 bc c2 bd  |................|
00000120  c2 be c2 bf c3 80 c3 81  c3 82 c3 83 c3 84 c3 85  |................|
00000130  c3 86 c3 87 c3 88 c3 89  c3 8a c3 8b c3 8c c3 8d  |................|
00000140  c3 8e c3 8f c3 90 c3 91  c3 92 c3 93 c3 94 c3 95  |................|
00000150  c3 96 c3 97 c3 98 c3 99  c3 9a c3 9b c3 9c c3 9d  |................|
00000160  c3 9e c3 9f c3 a0 c3 a1  c3 a2 c3 a3 c3 a4 c3 a5  |................|
00000170  c3 a6 c3 a7 c3 a8 c3 a9  c3 aa c3 ab c3 ac c3 ad  |................|
00000180  c3 ae c3 af c3 b0 c3 b1  c3 b2 c3 b3 c3 b4 c3 b5  |................|
00000190  c3 b6 c3 b7 c3 b8 c3 b9  c3 ba c3 bb c3 bc c3 bd  |................|
000001a0  c3 be c3 bf 5b 62 5d 0a  65 6e 64 0a              |....[b].end.|
000001ac


```

### Which characters to escape in link text

The output below provides hint to figure out.

```
{ echo -n '<a href="a.html">' ; recode -f </dev/null iso-8859-1/TEST8..utf-8 ; echo '"</a> end' ; } | pandoc  -i /dev/stdin -f html -t asciidoc | hd

00000000  6c 69 6e 6b 3a 61 2e 68  74 6d 6c 5b 00 01 02 03  |link:a.html[....|
00000010  04 05 06 07 08 20 0b 0c  20 0e 0f 10 11 12 13 14  |..... .. .......|
00000020  15 16 17 18 19 1a 1b 1c  1d 1e 1f 0a 21 22 23 24  |............!"#$|
00000030  25 26 27 28 29 2a 2b 2c  2d 2e 2f 30 31 32 33 34  |%&'()*+,-./01234|
00000040  35 36 37 38 39 3a 3b 3c  3d 3e 3f 40 41 42 43 44  |56789:;<=>?@ABCD|
00000050  45 46 47 48 49 4a 4b 4c  4d 4e 4f 50 51 52 53 54  |EFGHIJKLMNOPQRST|
00000060  55 56 57 58 59 5a 5b 5c  5d 5e 5f 60 61 62 63 64  |UVWXYZ[\]^_`abcd|
00000070  65 66 67 68 69 6a 6b 6c  6d 6e 6f 70 71 72 73 74  |efghijklmnopqrst|
00000080  75 76 77 78 79 7a 5c 7b  7c 7d 7e 7f e2 82 ac 3f  |uvwxyz\{|}~....?|
00000090  e2 80 9a c6 92 e2 80 9e  e2 80 a6 e2 80 a0 e2 80  |................|
000000a0  a1 cb 86 e2 80 b0 c5 a0  e2 80 b9 c5 92 3f c5 bd  |.............?..|
000000b0  3f 3f e2 80 98 e2 80 99  e2 80 9c e2 80 9d e2 80  |??..............|
000000c0  a2 e2 80 93 e2 80 94 cb  9c e2 84 a2 c5 a1 e2 80  |................|
000000d0  ba c5 93 3f c5 be c5 b8  c2 a0 c2 a1 c2 a2 c2 a3  |...?............|
000000e0  c2 a4 c2 a5 c2 a6 c2 a7  c2 a8 c2 a9 c2 aa c2 ab  |................|
000000f0  c2 ac c2 ad c2 ae c2 af  c2 b0 c2 b1 c2 b2 c2 b3  |................|
00000100  c2 b4 c2 b5 c2 b6 c2 b7  c2 b8 c2 b9 c2 ba c2 bb  |................|
00000110  c2 bc c2 bd c2 be c2 bf  c3 80 c3 81 c3 82 c3 83  |................|
00000120  c3 84 c3 85 c3 86 c3 87  c3 88 c3 89 c3 8a c3 8b  |................|
00000130  c3 8c c3 8d c3 8e c3 8f  c3 90 c3 91 c3 92 c3 93  |................|
00000140  c3 94 c3 95 c3 96 c3 97  c3 98 c3 99 c3 9a c3 9b  |................|
00000150  c3 9c c3 9d c3 9e c3 9f  c3 a0 c3 a1 c3 a2 c3 a3  |................|
00000160  c3 a4 c3 a5 c3 a6 c3 a7  c3 a8 c3 a9 c3 aa c3 ab  |................|
00000170  c3 ac c3 ad c3 ae c3 af  c3 b0 c3 b1 c3 b2 c3 b3  |................|
00000180  c3 b4 c3 b5 c3 b6 c3 b7  c3 b8 c3 b9 c3 ba c3 bb  |................|
00000190  c3 bc c3 bd c3 be c3 bf  22 5d 0a 65 6e 64 0a     |........"].end.|
0000019f
```

Tests:

```
echo "link:foo.html[a[b]c]" | asciidoctor - | grep foo.html

<p><a href="foo.html">a[b</a>c]</p>

echo "link:foo.html[a[b\]c]" | asciidoctor - | grep foo.html

<p><a href="foo.html">a[b]c</a></p>

```

**Comment by asamuzaK** [comment]
Thanks for the detail.

> * If lazy, use link:url[text] syntax always.

lazy, but seems best approach. :)

**Comment by asamuzaK** [comment]
How is IPv6 URL handled in asciidoc?
Could you tell me the result how will it be replaced in the examples below?

```
link:http://[::1]/[localhost]
link:http://\[::1\]/[localhost]
```

**Comment by fidergo-stephane-gourichon** [comment]
TL;DR: You raised an interesting head-scratcher. :-)

## Summary was not exactly true

What I wrote before...

> From the result below, I guess the set of characters to be escaped (percent-hexa url-encoded) in URL is:
> (...)
> 0x5B and 0x5D [ ]

...is true for the path part of the URL, yet it looks like the hostname/IP part should not have brackets escaped.

## Experimenting

The experiments below (section "details") aren't satisfying. The first two encodings are clearly mis-parsed.
The last encoding gets parsing right, but the generated URL is wrong.

## Conclusion 

IMHO asciidoc and asciidoctor should be fixed to parse this variant `link:http://\[::1\]/[localhost]` as a link with text "localhost" to URL `http://[::1]/`.

Best thing to do would be to discuss this on asciidoctor or asciidoc mailing-list.

In the meantime, what happens if you implement without considering this naked-IPV6-address case? You'll get something that works in other cases and that fits "release early, release often", right? If users need it they may contact you (that could be just me in some future).

## Details

```
for ENCODING in '[::1]' '\[::1\]' '%5B::1%5D' ; do echo ; echo ========================== ; echo "ENCODING $ENCODING" ; echo "link:http://$ENCODING/[localhost]" >test_ipv6_address.txt ; hd test_ipv6_address.txt ; for PARSER in asciidoc asciidoctor ; do echo ; echo With $PARSER ; $PARSER test_ipv6_address.txt ; grep localhost test_ipv6_address.html ; done ; done


==========================
ENCODING [::1]
00000000  6c 69 6e 6b 3a 68 74 74  70 3a 2f 2f 5b 3a 3a 31  |link:http://[::1|
00000010  5d 2f 5b 6c 6f 63 61 6c  68 6f 73 74 5d 0a        |]/[localhost].|
0000001e

With asciidoc
<div class="paragraph"><p><a href="http://">::1</a>/[localhost]</p></div>

With asciidoctor
<p><a href="http://">::1</a>/[localhost]</p>

==========================
ENCODING \[::1\]
00000000  6c 69 6e 6b 3a 68 74 74  70 3a 2f 2f 5c 5b 3a 3a  |link:http://\[::|
00000010  31 5c 5d 2f 5b 6c 6f 63  61 6c 68 6f 73 74 5d 0a  |1\]/[localhost].|
00000020

With asciidoc
<div class="paragraph"><p><a href="http://\">::1]/[localhost</a></p></div>

With asciidoctor
<p><a href="http://\">::1]/[localhost</a></p>

==========================
ENCODING %5B::1%5D
00000000  6c 69 6e 6b 3a 68 74 74  70 3a 2f 2f 25 35 42 3a  |link:http://%5B:|
00000010  3a 31 25 35 44 2f 5b 6c  6f 63 61 6c 68 6f 73 74  |:1%5D/[localhost|
00000020  5d 0a                                             |].|
00000022

With asciidoc
<div class="paragraph"><p><a href="http://%5B::1%5D/">localhost</a></p></div>

With asciidoctor
<p><a href="http://%5B::1%5D/">localhost</a></p>
```

**Comment by asamuzaK** [comment]
> In the meantime, what happens if you implement without considering this naked-IPV6-address case?

Maybe something like this,
```
const encodeSpecialChar = str => {
  const url = new URL(str);
  return url.href.replace(/([\s\[\]^`{|}])/g, c => encodeURIComponent(c));
};

encodeSpecialChar("http://example.com/q=[a b]"); // http://example.com/q=%5Ba%20b%5D
encodeSpecialChar("http://[::1]/"); // http://%5B::1%5D/
```

Since IPv6 has become widespread recently, I do not want to implement it like this although I know that there is an obvious bug.

**Comment by fidergo-stephane-gourichon** [comment]
An apparently well-founded answer seems to be on a very interesting article [How do you escape a complete URI? | Unspecified Behaviour](https://unspecified.wordpress.com/2012/02/12/how-do-you-escape-a-complete-uri/ "How do you escape a complete URI? | Unspecified Behaviour").
That article is more general than the topic here but interesting to avoid some pitfalls and understand why. 

(Incidentally, the bug where the url2clipboard menu item remains greyed out did appear again why trying to copy page + title as markdown link, and was worked around by reloading that page. firefox 57 on Ubuntu 16.04 AMD64).

> So what exactly should I do?
> 
> If you are building a URI programmatically, you must encode each component individually before composing them.

That page mentions a workaround on [encodeURI() - JavaScript | MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI#Description "encodeURI() - JavaScript | MDN")

> Also note that if one wishes to follow the more recent RFC3986 for URLs, which makes square brackets reserved (for IPv6) and thus not encoded when forming something which could be part of a URL (such as a host), the following code snippet may help:
> 
> function fixedEncodeURI(str) {
>     return encodeURI(str).replace(/%5B/g, '[').replace(/%5D/g, ']');
> }

But those global replace looks really suspicious.

## Conclusion

So, to be safe I think one must separate the URI string into two [parts](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier): (1) the scheme-//-authority, separated from (2) by the next slash, then (2) the remainder (path-query-fragment).
Then, keep scheme-//-authority as is, perform escaping of specific chars on that remainder, then concatenate.

This should be complete and safe. :-)

**Comment by asamuzaK** [comment]
Fixed.
Note that asciidoc has a bug when converting IPv6 URL to HTML anchor.
That is a matter of asciidoc, not this add-on.
Please contact asciidoc to fix this.

I'll release a test version later.

**Comment by asamuzaK** [comment]
Try build [v0.18.0a2](https://addons.mozilla.org/en-US/firefox/addon/url2clipboard/versions/beta)

**Comment by fidergo-stephane-gourichon** [comment]
Tested briefly, v0.18.0a2 is fine regarding this topic.

**Comment by asamuzaK** [comment]
Released [v0.18.0](https://addons.mozilla.org/en-US/firefox/addon/url2clipboard/ "Copy URL To Clipboard – Add-ons for Firefox")
Note that beta version will not update to release version.
You need to update manually.
